quixstreams/rowconsumer/consumer.py

```diff
@@ -225,5 +283,68 @@ def resume_backpressured(self):
 
     def reset_backpressure(self):
         # Reset the timeout back to its initial state
-        self._backpressure_resume_at = _MAX_FLOAT
+        self._backpressure_resume_at = float("inf")
         self._backpressurred_tps.clear()
+
+    def poll_buffered(
+        self, timeout: Optional[float] = None
+    ) -> Optional[SuccessfulConfluentKafkaMessageProto]:
+        """
+        Poll messages in a buffered way to provide in-order reads across multiple
+        topic partitions with the same partition number.
+        """
+
+        # Probe the buffer and return immediately if there's data available
+        msg = self._buffer.pop()
+        if msg is None:
+            # If the buffer is empty, feed it and try the probing again
+            self._feed_buffer(timeout=timeout)
+            msg = self._buffer.pop()
+
+        return msg
+
+    def _feed_buffer(self, timeout: Optional[float] = None):
+        """
+        Feed the internal buffer and pause or resume the assigned partitions for the
+        balanced consumption.
+
+        :param timeout: the maximum time in seconds to wait for the next batch of messages.
+        """
+        try:
+            messages = self.consume(
+                num_messages=self._max_partition_buffer_size, timeout=timeout
+            )
+        except PartitionAssignmentError:
+            # Always propagate errors happened during assignment
+            raise
+        except Exception as exc:
+            to_suppress = self._on_error(exc, None, logger)
+            if not to_suppress:
+                raise
+            messages = []
+
+        # Get the recent cached high watermarks
+        high_watermarks: dict[tuple[str, int], int] = {}
+        for tp in self.assignment():
+            topic_obj = self._topics[tp.topic]
+            if not topic_obj.is_changelog:
+                _, high = self.get_watermark_offsets(partition=tp, cached=True)
+                high_watermarks[(tp.topic, tp.partition)] = high
```

Comment:
Not sure how this would affect the performance, but from a readability perspective, this may be done as `TopicPartition` is a hashable object.
```suggestion
                high_watermarks[tp] = high
```

Reply:
I prefer to keep it this way to avoid using `confluent_kafka.TopicPartition` in the `PartitionBufferGroup` class too.